core/src/nats.rs

```diff
@@ -300,24 +296,46 @@ impl TypedNats {
 
         let consumer = stream
             .create_consumer(async_nats::jetstream::consumer::pull::Config {
-                deliver_policy: DeliverPolicy::Last,
+                deliver_policy,
                 filter_subject: subject.subject.clone(),
                 ..async_nats::jetstream::consumer::pull::Config::default()
             })
             .await
             .to_anyhow()?;
 
-        let result = match timeout(
-            Duration::from_secs(1),
-            consumer.stream().messages().await.to_anyhow()?.next(),
-        )
-        .await
-        {
-            Ok(Some(v)) => v.to_anyhow()?,
-            _ => return Ok(None),
-        };
-
-        Ok(Some(serde_json::from_slice(&result.payload)?))
+        let mut result: Vec<T> = Vec::new();
+
+        loop {
+            let mut messages = consumer.fetch().messages().await.to_anyhow()?;
+            let mut done = true;
+
+            while let Some(v) = messages.next().await {
+                let v = v.to_anyhow()?;
+                done = false;
+
+                result.push(serde_json::from_slice(&v.payload)?);
+
+                if v.info().to_anyhow()?.pending == 0 {
```

Comment:
so pending is the number of messages the server thinks haven't been .next()ed yet right? or is it messages that the server has but the client hasn't got in its queue yet?

guess what I'm asking is, is it possible for pending to be 0 while messages.next() != None

Reply:
Pending is the number remaining not counting the current message. Unfortunately, when no messages remain, `messages.next().await` returns a `Some(Err(_))` rather than `None`. Because `async_nats` uses boxed errors, and the error itself only indicates the type of error (no messages on server) through a string, checking the error message directly would be pretty ugly. The approach here is to try to avoid encountering the error in the first place, by only calling `.next()` a second time if we know there are messages.

That said, looking at this again, I think it will cause an undesired errors if there are no messages at all on the server, since it will never have a pending value to read. Ugh. I'll revisit this, good catch.

Reply:
:/, thats ew, could you take_while() pending != 0, and do the map_err thing to stop reading if there's any errors?

Reply:
like `messages.take_while(async |res| res.ok_or(false).map(|msg| msg.info().pending > 0)` with whatever prayers it takes for the compiler to accept it.

Reply:
The problem is I don't want to treat any old error the same as the error that happens when there are not messages in the stream. My hack for now is to string match to see if the error is the right one (since async_nats doesn't yet have concrete errors, this is the only way.)